articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,307 @@
+Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory.
+This is reminiscent of the pattern used earlier Keymint [1].
+
+
+```
+$ tree keystore/
+keystore
+├── contexts
+│   └── ...
+│       └── ...
+├── private
+│   ├── ca.csr.pem
+│   └── ca.key.pem
+└── public
+    ├── ca.cert.pem
+    ├── identity_ca.cert.pem
+    └── permissions_ca.cert.pem
+```
+
+
+### ``public``
+
+The ``public`` directory contains anything permissable as public, such as public certificates for the identity or permissions certificate authorities.
+As such, this can be given read access to all executables.
+Note that in the default case, both the `identity_ca` and `permissions_ca` are the same CA certificate.
+
+### ``private``
+
+The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
+This directory should be redacted before deploying the keystore onto the target device/robot.
+
+### ``contexts``
+
+The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
+Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
+
+
+## Runtime
+
+TODO: Some transition paragraph here about ros launch
+
+### Unqualified context path
+
+For nodes with unqualified context paths, the context directory will subsequently default to the root level context.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <node pkg="demo_nodes_cpp" exec="listener"/>
+</launch>
+```
+
+```
+$ tree contexts/
+contexts/
+├── cert.pem
+├── governance.p7s
+├── identity_ca.cert.pem
+├── key.pem
+├── permissions_ca.cert.pem
+└── permissions.p7s
+```
+
+### Pushed unqualified context path
+
+For nodes with unqualified context paths pushed by a namespace, the context directory will subsequently be pushed to the relative sub-folder.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener"/>
+  </group>
+</launch>
+```
+
+```
+$ tree --dirsfirst contexts/
+contexts/
+├── foo
+│   ├── cert.pem
+│   ├── governance.p7s
+│   ├── identity_ca.cert.pem
+│   ├── key.pem
+│   ├── permissions_ca.cert.pem
+│   └── permissions.p7s
+├── cert.pem
+├── governance.p7s
+├── identity_ca.cert.pem
+├── key.pem
+├── permissions_ca.cert.pem
+└── permissions.p7s
+```
+
+### Relatively pushed qualified context path
+
+For nodes with qualified context paths pushed by a namespace, the qualified context directory will subsequently be pushed to the relative sub-folder.
+
+``` xml
+<launch>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener" context="bar"/>
```

Comment:
I would use `context_name` instead of just `context`.

Reply:
I'm not sure what to call it, as it could be a _path or just a relative _name, so I left it as context.

Reply:
If we left only one attribute for specifying the context name and not two, using `context` is fine.